File: SwiftUI/EdgeStudio/Components/ProfileViewer/PlanNodeBox.swift
import SwiftUI

/// Single rounded box for the Plan tree view. Distinct from
/// `ProfileOperatorCard` (used by Card mode): smaller, fixed-width,
/// and color-coded by hotspot status so an operator burning a
/// disproportionate share of the request's elapsed time pops visually.
///
/// Visual hierarchy (top-to-bottom inside the box):
///   1. **Operator name** (bold, monospaced) — `"scan"`, `"limit"`,
///      `"sequence"`, …
///   2. **Key attribute** (subtle) — best-effort distinguishing
///      attribute: `collection` for scans, `limit` for limits, etc.
///      Hidden for operators without one.
///   3. **Total time** — `exec + recv + send` rendered through
///      `ProfileTimeFormatter`, with an optional `(N.N%)` suffix
///      when the node represents ≥ 5% of the total elapsed time.
///   4. **In/out doc counts** — `"N in / M out"`, with both halves
///      hidden if the operator emits neither.
///
/// Colors:
///   - **Default**: muted green — matches the reference layout in
///     `screens/couchbase-plan.png` and signals "normal".
///   - **Hotspot**: orange when this operator's `exec` exceeds
///     `hotspotThreshold` (50%) of the request's total `elapsedNs`.
///     Same threshold that powers the percent badge — see
///     `plans/dql-profile-feature.md` "Bottleneck threshold".
struct PlanNodeBox: View {
    let node: QueryProfileOperator
    let totalElapsedNs: Int64

    /// Operator's exec time represents > 50% of total elapsed → flag
    /// as a bottleneck. Threshold lives here so the orange tint and
    /// the percent badge agree.
    static let hotspotThreshold = 0.5

    private var isHotspot: Bool {
        guard let execNs = node.stats?.execNs, totalElapsedNs > 0 else { return false }
        return Double(execNs) / Double(totalElapsedNs) >= Self.hotspotThreshold
    }

    private var fillColor: Color {
        isHotspot ? Color.orange.opacity(0.18) : Color.green.opacity(0.16)
    }

    private var borderColor: Color {
        isHotspot ? Color.orange.opacity(0.5) : Color.green.opacity(0.45)
    }

    /// Best-effort identifying attribute for the node title row.
    /// We surface the first attribute that names *what* the operator
    /// works on rather than internal config (`descriptor` /
    /// `datasource` add noise without help). Falls back to nil — the
    /// box just omits the subtitle line.
    private var keyAttribute: String? {
        let preferred = ["collection", "alias", "limit", "field", "table"]
        for key in preferred {
            if let attr = node.attributes.first(where: { $0.key == key }) {
                return attr.value
            }
        }
        return nil
    }

    /// Sum of the non-overlapping phases. Ditto's per-operator
    /// `exec`/`recv`/`send` are disjoint phases on the operator's
    /// own thread — adding them gives the wall-time the operator
    /// occupied without double-counting.
    private var totalNs: Int64? {
        guard let stats = node.stats else { return nil }
        let parts: [Int64] = [stats.execNs, stats.recvNs, stats.sendNs].compactMap(\.self)
        guard !parts.isEmpty else { return nil }
        return parts.reduce(0, +)
    }

    private var timeLabel: String? {
        guard let totalNs else { return nil }
        var label = ProfileTimeFormatter.format(ns: totalNs)
        if let pct = ProfileTimeFormatter.percentOfTotal(ns: totalNs, totalNs: totalElapsedNs) {
            label += "  (\(pct))"
        }
        return label
    }

    private var ioLabel: String? {
        let input = node.stats?.documentsIn
        let output = node.stats?.documentsOut
        switch (input, output) {
        case let (input?, output?):
            return "\(ProfileFormat.documents(input)) in / \(ProfileFormat.documents(output)) out"
        case (nil, let output?):
            return "\(ProfileFormat.documents(output)) out"
        case (let input?, nil):
            return "\(ProfileFormat.documents(input)) in"
        default:
            return nil
        }
    }

    var body: some View {
        VStack(alignment: .center, spacing: 4) {
            Text(node.name)
                .font(.subheadline.weight(.semibold).monospaced())
                .foregroundStyle(.primary)
            if let keyAttribute {
                Text(keyAttribute)
                    .font(.caption.monospaced())
                    .foregroundStyle(.secondary)
                    .lineLimit(1)
                    .truncationMode(.middle)
            }
            if let timeLabel {
                Text(timeLabel)
                    .font(.caption.monospacedDigit())
                    .foregroundStyle(.primary.opacity(0.85))
            }
            if let ioLabel {
                Text(ioLabel)
                    .font(.caption2.monospacedDigit())
                    .foregroundStyle(.secondary)
            }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        .frame(width: 160)
        .background(
            RoundedRectangle(cornerRadius: 10)
                .fill(fillColor)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 10)
                .strokeBorder(borderColor, lineWidth: 1)
        )
        // Subtle elevation cue — matches the "card with depth" feel
        // of the reference Couchbase screenshot without going overboard
        // with shadow that would clash against dark mode.
        .shadow(color: Color.black.opacity(0.08), radius: 2, x: 0, y: 1)
    }
}

File: android/app/src/main/java/com/costoda/dittoedgestudio/ui/mainstudio/profile/PlanNodeBox.kt
package com.costoda.dittoedgestudio.ui.mainstudio.profile

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.costoda.dittoedgestudio.data.repository.ProfileTimeFormatter
import com.costoda.dittoedgestudio.domain.model.QueryProfileOperator
import kotlin.math.max

/**
 * Single operator node in the plan-tree view. Renders the operator name, exec time,
 * and a percentage share badge (`execNs / totalExecNs`). Hotspot color (errorContainer)
 * when the operator's share is >= 50% of the whole plan's exec time.
 */
@Composable
fun PlanNodeBox(
    operator: QueryProfileOperator,
    totalExecNs: Long,
    modifier: Modifier = Modifier,
) {
    val ownExec = operator.stats?.execNs ?: 0L
    val share = if (totalExecNs > 0L) (ownExec * 100L) / max(totalExecNs, 1L) else 0L
    val hotspot = share >= 50L
    val containerColor = if (hotspot) {
        MaterialTheme.colorScheme.errorContainer
    } else {
        MaterialTheme.colorScheme.surfaceContainerHigh
    }
    val contentColor = if (hotspot) {
        MaterialTheme.colorScheme.onErrorContainer
    } else {
        MaterialTheme.colorScheme.onSurface
    }

    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(8.dp),
        color = containerColor,
        contentColor = contentColor,
    ) {
        Column(modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = operator.name,
                    style = MaterialTheme.typography.titleSmall,
                    modifier = Modifier,
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    text = "$share%",
                    style = MaterialTheme.typography.labelMedium,
                )
            }
            if (ownExec > 0L) {
                Text(
                    text = "exec: ${ProfileTimeFormatter.format(ownExec)}",
                    style = MaterialTheme.typography.bodySmall,
                )
            }
        }
    }
}
